quantum/pointing_device/pointing_device_modes.c

```diff
@@ -0,0 +1,674 @@
+    int16_t count = 0;
+    uint8_t dir   = current_pointing_mode_direction();
+
+    switch (dir) {
+        case PD_DOWN ... PD_UP:
+            count = divisor_divide16(pointing_modes[current_device].y);
+            if (!count) return; // exit if accumulated y is too low
+            pointing_modes[current_device].y -= divisor_multiply16(count);
+            pointing_modes[current_device].x = 0;
+            break;
+        case PD_LEFT ... PD_RIGHT:
+            count = divisor_divide16(pointing_modes[current_device].x);
+            if (!count) return; // exit if accumulated x is too low
+            pointing_modes[current_device].x -= divisor_multiply16(count);
+            pointing_modes[current_device].y = 0;
+            break;
+    }
+    // skip if KC_TRNS or KC_NO (but allow for axes update above)
+    //if (pm_keycodes[dir] < 2) return;
+
+    // tap codes
+    uint8_t taps = clamp_uint_16_to_8((uint16_t)abs(count));
+    do {
+        tap_code16_delay(pm_keycodes[dir], POINTING_TAP_DELAY);
+    } while (--taps);
+}
+
+/**
+ * @brief external wrapper for pointing_tap_keycodes to allow easier input
+ *
+ * keycode order follow VIM convention (LEFT, DOWN, UP, RIGHT).s
+ *
+ * @params kc_left[in]  uint16_t keycode for negative x
+ * @params kc_down[in]  uint16_t keycode for negative y
+ * @params kc_up[in]    uint16_t keycode for positive y
+ * @params kc_right[in] uint16_t keycode for positive x
+ */
+void pointing_tap_codes(uint16_t kc_left, uint16_t kc_down, uint16_t kc_up, uint16_t kc_right) {
+    uint16_t pm_keycodes[4] = {kc_down, kc_up, kc_left, kc_right};
+    pointing_tap_keycodes_raw(pm_keycodes);
+}
+
+/**
+ * @brief Tap keycodes from pointing mode maps
+ *
+ * Will translate internal x & y axes into mode map keycode taps.
+ * and will update the x and y values.-
+ *
+ * NOTE: favours staying on axis and favours the horizontal over the vertical axis
+ *       and will clear the orthogonal axis
+ *
+ * @params map_id id of current map
+ */
+static void pointing_exec_mapping(uint8_t map_id) {
+    int16_t count = 0;
+    uint8_t dir   = current_pointing_mode_direction();
+
+    switch (dir) {
+        case PD_DOWN ... PD_UP:
+            count = divisor_divide16(pointing_modes[current_device].y);
+            if (!count) return; // exit if accumulated y is too low
+            pointing_modes[current_device].y -= divisor_multiply16(count);
+            pointing_modes[current_device].x = 0;
+            break;
+        case PD_LEFT ... PD_RIGHT:
+            count = divisor_divide16(pointing_modes[current_device].x);
+            if (!count) return; // exit if accumulated x is too low
+            pointing_modes[current_device].x -= divisor_multiply16(count);
+            pointing_modes[current_device].y = 0;
+            break;
+    }
+
+    // tap codes
+    uint8_t taps = clamp_uint_16_to_8((uint16_t)abs(count));
+    do {
+        action_exec(MAKE_POINTING_MODE_EVENT(map_id, dir, true));
```

Comment:
> quantum/pointing_device/pointing_device_modes.c:485:21: error: implicit declaration of function 'MAKE_POINTING_MODE_EVENT' [-Werror=implicit-function-declaration]

Looks like this needs some preprocessor guarding

Reply:
Ah I see yup that shouldn't go through without `#define POINTING_MODE_MAP_ENABLE` in config.h